### PR TITLE
Migrate to kubebuilder v4

### LINKED ISSUE #365: Migrate nifikop to KubeBuilder go/v4 from go/v3
### Is your feature request related to a problem?

No

### Describe the solution you'd like to see

KubeBuilder go/v4 is now stable as of operator-sdk v1.33.0, so we should migrate nifikop to leverage all of the new features detailed here: https://book.kubebuilder.io/migration/v3vsv4

Here are the migration instructions: https://book.kubebuilder.io/migration/migration_guide_gov3_to_gov4

If there are no CRD breaking changes, this can safely be done in nifikop v1.x. Otherwise, we should target the Release 2.0.0 milestone.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mh013370**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | no
| API breaks?     | no
| Deprecations?   | no
| Related tickets | fixes #365 
| License         | Apache 2.0

### What's in this PR?

Migrates the operator to KubeBuilder go/v4 and operator-sdk v1.33.0, applying all appropriate migrations in the process. The migration process required re-scaffolding the entire API and all controllers in order to regenerate the `PROJECT` configuration, but the actual specs & controller code was copied forward. Regardless, care should be taken in reviewing this to ensure that i did not incidentally leave something out.

I updated the `Makefile` and all generated Kustomize configuration to match the latest in KubeBuilder. From this point, only minor updates should be required.

Primary changes:
- `main.go` moved to `cmd/main.go`
- `controllers/` moved to `internal/controller/`
- All of the `config/` generated yaml was updated

**Note** Merging this will necessitate updating your locally installed `operator-sdk` to v1.33.0.

### Why?

KubeBuilder go/v4 is now stable and go/v3 will be end-of-lifed and removed.

### Checklist

- [ ] Implementation tested
- [x] Error handling code meets the [guideline](docs/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)
- [x] Append changelog with changes

### CODE REVIEW

**Review by juldrixx — CHANGES_REQUESTED** [review]
Before taking an in-depth look and testing it, I think that the version of Operator SDK is mentionned in the documention (not sure for Kubebuilder itself), so it must be updated accordingly.

**Comment by mh013370** [comment]
> Before taking an in-depth look and testing it, I think that the version of Operator SDK is mentionned in the documention (not sure for Kubebuilder itself), so it must be updated accordingly.

Yeah, so basically upgrading to operador-sdk v1.33.0 is just migrating to kubebuilder go/v4: 

https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.33.0/

**Review by r65535 — COMMENTED** [review]
Initial checks LGTM.
I'll test a few cluster deployments using this branch tomorrow 👍

**Comment by juldrixx** [comment]
> > Before taking an in-depth look and testing it, I think that the version of Operator SDK is mentionned in the documention (not sure for Kubebuilder itself), so it must be updated accordingly.
> 
> Yeah, so basically upgrading to operador-sdk v1.33.0 is just migrating to kubebuilder go/v4:
> 
> https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.33.0/

I was thinking of NiFiKop documentation 😄

**Comment by mh013370** [comment]
> ng of NiFiKop document



> > > Before taking an in-depth look and testing it, I think that the version of Operator SDK is mentionned in the documention (not sure for Kubebuilder itself), so it must be updated accordingly.
> > 
> > 
> > Yeah, so basically upgrading to operador-sdk v1.33.0 is just migrating to kubebuilder go/v4:
> > https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.33.0/
> 
> I was thinking of NiFiKop documentation 😄

You're right it is. Updated :)

File: go.mod [hunk 1]
```diff
@@ -1,108 +1,111 @@
 module github.com/konpyutaika/nifikop
 
-go 1.21
+go 1.21.5
```
[the remark this answers is not in the document]
Comment (mh013370) [reply go.mod:3]: KubeBuilder actually added the minor version, but i like that it's explicit so when `govuln` finds an issue prompting an update we can just go in an up the minor version.

Thoughts?

[the remark this answers is not in the document]
Comment (r65535) [reply go.mod:3]: Yeah I agree - it requires more maintenance from committers but think it's worth the price.
It's also explicit for users, so they can see which version we're using for each release.

Comment (juldrixx) [inline go.mod:3]:
```suggestion
go 1.21.7
```

**Review by juldrixx — CHANGES_REQUESTED** [review]

**Review by juldrixx — APPROVED** [review]

**Pull request merged**